pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -0,0 +1,365 @@
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/sync/errgroup"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storage/sharding"
+	"github.com/grafana/mimir/pkg/util"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+const encodingTypeSnappyFramed = "x-snappy-framed"
+
+var errShardCountTooLow = errors.New("shard count too low")
+
+type shardActiveSeriesMiddleware struct {
+	upstream http.RoundTripper
+	limits   Limits
+	logger   log.Logger
+	encoder  *s2.Writer
+}
+
+func newShardActiveSeriesMiddleware(upstream http.RoundTripper, limits Limits, logger log.Logger) http.RoundTripper {
+	return &shardActiveSeriesMiddleware{
+		upstream: upstream,
+		limits:   limits,
+		logger:   logger,
+		encoder:  s2.NewWriter(nil),
+	}
+}
+
+func (s *shardActiveSeriesMiddleware) RoundTrip(r *http.Request) (*http.Response, error) {
+	spanLog, ctx := spanlogger.NewWithLogger(r.Context(), s.logger, "shardActiveSeries.RoundTrip")
+	defer spanLog.Finish()
+
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	defaultShardCount := s.limits.QueryShardingTotalShards(tenantID)
+	shardCount := setShardCountFromHeader(defaultShardCount, r, spanLog)
+
+	if shardCount < 2 {
+		spanLog.DebugLog("msg", "query sharding disabled for request")
+		return s.upstream.RoundTrip(r)
+	}
+
+	if maxShards := s.limits.QueryShardingMaxShardedQueries(tenantID); shardCount > maxShards {
+		return nil, apierror.New(
+			apierror.TypeBadData,
+			fmt.Sprintf("shard count %d exceeds allowed maximum (%d)", shardCount, maxShards),
+		)
+	}
+
+	selector, err := parseSelector(r)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	spanLog.DebugLog(
+		"msg", "sharding active series query",
+		"shardCount", shardCount, "selector", selector.String(),
+	)
+
+	reqs, err := buildShardedRequests(ctx, r, shardCount, selector)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeInternal, err.Error())
+	}
+
+	resp, err := doShardedRequests(ctx, reqs, s.upstream)
+	if err != nil {
+		if errors.Is(err, errShardCountTooLow) {
+			return nil, apierror.New(apierror.TypeBadData, fmt.Errorf("%w: try increasing the requested shard count", err).Error())
+		}
+		return nil, apierror.New(apierror.TypeInternal, err.Error())
+	}
+
+	return s.mergeResponses(ctx, resp, r.Header.Get("Accept-Encoding")), nil
+}
+
+func setShardCountFromHeader(origShardCount int, r *http.Request, spanLog *spanlogger.SpanLogger) int {
+	for _, value := range r.Header.Values(totalShardsControlHeader) {
+		shards, err := strconv.ParseInt(value, 10, 32)
+		if err != nil {
+			continue
+		}
+		if shards >= 0 {
+			spanLog.DebugLog(
+				"msg",
+				fmt.Sprintf("using shard count from header %s: %d", totalShardsControlHeader, shards),
+			)
```

Comment:
Nit, but there's at least one unnecessary line break here, and IMO all of them are unnecessary.
```suggestion
			spanLog.DebugLog("msg", fmt.Sprintf("using shard count from header %s: %d", totalShardsControlHeader, shards))
```